In [2]:
import numpy as np
import pandas as pd
import time

In [3]:
arr1=np.random.rand(100,100)
arr2=np.random.rand(100,100)
arr1
arr2

array([[0.94655893, 0.99389658, 0.13206096, ..., 0.32960313, 0.01162421,
        0.54476093],
       [0.56251253, 0.34345693, 0.20907943, ..., 0.82966404, 0.03326191,
        0.60989543],
       [0.8019083 , 0.02794786, 0.65258113, ..., 0.23668012, 0.13498012,
        0.79331185],
       ...,
       [0.4539753 , 0.09591007, 0.78015434, ..., 0.8449376 , 0.64254553,
        0.70581081],
       [0.56985264, 0.85382708, 0.61708154, ..., 0.81239402, 0.3015387 ,
        0.35317119],
       [0.95897091, 0.53392056, 0.67813513, ..., 0.45212587, 0.98954518,
        0.53807645]])

array([[0.2984269 , 0.72526244, 0.191573  , ..., 0.41454467, 0.42033238,
        0.75104407],
       [0.7616134 , 0.54080152, 0.99102574, ..., 0.83004538, 0.21835991,
        0.09425773],
       [0.24581988, 0.62414265, 0.72840187, ..., 0.54418784, 0.68059433,
        0.71745236],
       ...,
       [0.38507781, 0.71232684, 0.23935094, ..., 0.53309901, 0.25058979,
        0.84344133],
       [0.72374104, 0.10174211, 0.96533408, ..., 0.04665998, 0.46883947,
        0.26444206],
       [0.59290043, 0.38217471, 0.84583819, ..., 0.42124934, 0.36814835,
        0.19069052]])

In [4]:
arr3=np.zeros(100*100).reshape(100,100)
start=time.clock()
for i in range(100):
    for j in range(100):
        arr3[i][j]=arr1[i][j]*arr2[i][j]
end=time.clock()
arr3
end-start

C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
  
C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:6: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
  


array([[0.28247864, 0.72083586, 0.02529931, ..., 0.13663522, 0.00488603,
        0.40913946],
       [0.42841708, 0.18574203, 0.2072031 , ..., 0.6886588 , 0.00726307,
        0.05748736],
       [0.197125  , 0.01744345, 0.47534132, ..., 0.12879844, 0.09186671,
        0.56916346],
       ...,
       [0.17481582, 0.06831932, 0.18673067, ..., 0.4504354 , 0.16101535,
        0.59531001],
       [0.41242575, 0.08687017, 0.59568983, ..., 0.03790629, 0.14137325,
        0.09339331],
       [0.56857427, 0.20405094, 0.57359259, ..., 0.19045772, 0.36429942,
        0.10260608]])

0.09705870000000516

In [5]:
start=time.clock()
arr4=arr1*arr2
end=time.clock()
arr4
end-start

C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
  """Entry point for launching an IPython kernel.
C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
  This is separate from the ipykernel package so we can avoid doing imports until


array([[0.28247864, 0.72083586, 0.02529931, ..., 0.13663522, 0.00488603,
        0.40913946],
       [0.42841708, 0.18574203, 0.2072031 , ..., 0.6886588 , 0.00726307,
        0.05748736],
       [0.197125  , 0.01744345, 0.47534132, ..., 0.12879844, 0.09186671,
        0.56916346],
       ...,
       [0.17481582, 0.06831932, 0.18673067, ..., 0.4504354 , 0.16101535,
        0.59531001],
       [0.41242575, 0.08687017, 0.59568983, ..., 0.03790629, 0.14137325,
        0.09339331],
       [0.56857427, 0.20405094, 0.57359259, ..., 0.19045772, 0.36429942,
        0.10260608]])

0.0017162999999982276

#### **Obivously,the second method is better.

In [7]:
arr5=np.random.rand(10086,2020)
arr6=np.random.rand(10086,2020)
arr7=np.zeros(10086*2020).reshape(10086,2020)
start=time.clock()
for i in range(10086):
    for j in range(2020):
        arr7[i][j]=arr5[i][j]*arr6[i][j]
end=time.clock()
arr7
print('method 1 costs',end-start)
start=time.clock()
arr7=arr5*arr6
end=time.clock()
print('method 2 costs',end-start)

C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
  after removing the cwd from sys.path.
C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:8: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
  


array([[0.51143197, 0.16949964, 0.02273143, ..., 0.03831336, 0.62594391,
        0.23708622],
       [0.38957571, 0.01036969, 0.10196881, ..., 0.35483479, 0.5119051 ,
        0.13553017],
       [0.0821058 , 0.00446635, 0.18453127, ..., 0.46509007, 0.08013066,
        0.15091133],
       ...,
       [0.65519906, 0.01890354, 0.11352919, ..., 0.02889959, 0.01386876,
        0.20744745],
       [0.53990499, 0.02275615, 0.25511664, ..., 0.73926309, 0.4143961 ,
        0.14043719],
       [0.34638312, 0.00353081, 0.12144538, ..., 0.09245535, 0.35289834,
        0.62268712]])

C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:11: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
  # This is added back by InteractiveShellApp.init_path()


method 1 costs 61.766058900000004
method 2 costs 0.2960581000000104


C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:13: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
  del sys.path[0]


In [8]:
arr1=pd.DataFrame(arr1)
list1=np.random.rand(100*100).reshape(1,100*100)
list1=list(list1)
list1

[array([0.16670507, 0.24959901, 0.5057495 , ..., 0.29268343, 0.11927646,
        0.58189279])]

In [11]:
A=np.zeros(100*100).reshape(100,100)
# Method 1: A is an array
start=time.clock()
for i in range(100):
    for j in range(100):
        A[i, j] = i * 100 + j
end=time.clock()
print('Method 1 costs',end-start)
# Method 2: B is a DataFrame
B=pd.DataFrame(np.zeros(100*100).reshape(100,100))
start=time.clock()
for i in range(100):
    for j in range(100):
        B.iloc[i, j] = i * 100 + j
end=time.clock()
print('Method 2 costs',end-start)
# Method 3: C is a DataFrame
C=pd.DataFrame(np.zeros(100*100).reshape(100,100))
start=time.clock()
for i in range(100):
    Temp = np.array([i * 100 + j for j in range(100)])
    C.iloc[i, :] = Temp
end=time.clock()
print('Method 3 costs',end-start)
# Method 4: D is a DataFrame
D=pd.DataFrame(np.zeros(100*100).reshape(100,100))
start=time.clock()
for j in range(100):
    Temp = np.array([i * 100 + j for i in range(100)])
    D.iloc[:, j] = Temp
end=time.clock()
print('Method 4 costs',end-start)
# Method 5: E is a list
E=[0 for i in range(100*100)]
start=time.clock()
for i in range(100):
    for j in range(100):
        E[i * 100 + j] = i * 100 + j
E = np.array(A).reshape(100,100)
end=time.clock()
print('Method 5 costs',end-start)
# Method 6: F is a list
import numba
start=time.clock()
@numba.jit()#numba can accelerate the loop
def Test(F):
    for i in range(100):
        for j in range(100):
            F[i * 100 + j] = i * 100 + j
    return F
F = [0 for i in range(100*100)]
F = Test(F)
F = np.array(F).reshape(100,100)
end=time.clock()
print('Method 6 costs',end-start)

C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:7: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
  import sys
C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:11: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
  # This is added back by InteractiveShellApp.init_path()


Method 1 costs 0.017564999999990505


C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:15: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
  from ipykernel import kernelapp as app
C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:19: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:23: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:27: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead


Method 2 costs 4.567620499999975
Method 3 costs 0.058825899999987996


C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:31: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:35: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:40: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead


Method 4 costs 0.16235449999999219
Method 5 costs 0.012785699999994904


C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:44: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
C:\Users\rcf\Anaconda3\lib\site-packages\numba\ir_utils.py:1958: NumbaPendingDeprecationWarning: 
Encountered the use of a type that is scheduled for deprecation: type 'reflected list' found for argument 'F' of function 'Test'.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-reflection-for-list-and-set-types

File "<ipython-input-11-de54a545463f>", line 46:
@numba.jit()#numba can accelerate the loop
def Test(F):
^

  warnings.warn(NumbaPendingDeprecationWarning(msg, loc=loc))


Method 6 costs 1.2799431000000254


C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:54: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead


Method 1 costs 0.017564999999990505
Method 2 costs 4.567620499999975
Method 3 costs 0.058825899999987996
Method 4 costs 0.16235449999999219
Method 5 costs 0.012785699999994904
Method 6 costs 1.2799431000000254
It is clear that method 2 costs the most time. The reason is that DataFrame follows column-major order, so every time we circularly assign the value to elements by row, we jump to another unit of this DataFrame once which definitely costs much more time than successively assigning the value to elements in the same unit.
Notice that Method 6 costs a lot of time as well, because defining a new function costs a lot.

In [13]:
# Chunk 1
A = np.array([i for i in range(10000)]).reshape(100, 100)
for i in A:
    A[i] = np.array([i for i in range(100)])
A
# Chunk 2
A = np.array([i for i in range(10000)]).reshape(100, 100)
A = pd.DataFrame(A)
for i in A:
    A[i] = np.array([i for i in range(100)])
A

array([[ 0,  1,  2, ..., 97, 98, 99],
       [ 0,  1,  2, ..., 97, 98, 99],
       [ 0,  1,  2, ..., 97, 98, 99],
       ...,
       [ 0,  1,  2, ..., 97, 98, 99],
       [ 0,  1,  2, ..., 97, 98, 99],
       [ 0,  1,  2, ..., 97, 98, 99]])

0   1   2   3   4   5   6   7   8   9   ...  90  91  92  93  94  95  96  \
0    0   0   0   0   0   0   0   0   0   0  ...   0   0   0   0   0   0   0   
1    1   1   1   1   1   1   1   1   1   1  ...   1   1   1   1   1   1   1   
2    2   2   2   2   2   2   2   2   2   2  ...   2   2   2   2   2   2   2   
3    3   3   3   3   3   3   3   3   3   3  ...   3   3   3   3   3   3   3   
4    4   4   4   4   4   4   4   4   4   4  ...   4   4   4   4   4   4   4   
5    5   5   5   5   5   5   5   5   5   5  ...   5   5   5   5   5   5   5   
6    6   6   6   6   6   6   6   6   6   6  ...   6   6   6   6   6   6   6   
7    7   7   7   7   7   7   7   7   7   7  ...   7   7   7   7   7   7   7   
8    8   8   8   8   8   8   8   8   8   8  ...   8   8   8   8   8   8   8   
9    9   9   9   9   9   9   9   9   9   9  ...   9   9   9   9   9   9   9   
10  10  10  10  10  10  10  10  10  10  10  ...  10  10  10  10  10  10  10   
11  11  11  11  11  11  11  11  11  11  11  ...  11  11  11  11  11  11  11   
12  12  12  12  12  12  12  12  12  12  12  ...  12  12  12  12  12  12  12   
13  13  13  13  13  13  13  13  13  13  13  ...  13  13  13  13  13  13  13   
14  14  14  14  14  14  14  14  14  14  14  ...  14  14  14  14  14  14  14   
15  15  15  15  15  15  15  15  15  15  15  ...  15  15  15  15  15  15  15   
16  16  16  16  16  16  16  16  16  16  16  ...  16  16  16  16  16  16  16   
17  17  17  17  17  17  17  17  17  17  17  ...  17  17  17  17  17  17  17   
18  18  18  18  18  18  18  18  18  18  18  ...  18  18  18  18  18  18  18   
19  19  19  19  19  19  19  19  19  19  19  ...  19  19  19  19  19  19  19   
20  20  20  20  20  20  20  20  20  20  20  ...  20  20  20  20  20  20  20   
21  21  21  21  21  21  21  21  21  21  21  ...  21  21  21  21  21  21  21   
22  22  22  22  22  22  22  22  22  22  22  ...  22  22  22  22  22  22  22   
23  23  23  23  23  23  23  23  23  23  23  ...  23  23  23  23  23  23  23   
24  24  24  24  24  24  24  24  24  24  24  ...  24  24  24  24  24  24  24   
25  25  25  25  25  25  25  25  25  25  25  ...  25  25  25  25  25  25  25   
26  26  26  26  26  26  26  26  26  26  26  ...  26  26  26  26  26  26  26   
27  27  27  27  27  27  27  27  27  27  27  ...  27  27  27  27  27  27  27   
28  28  28  28  28  28  28  28  28  28  28  ...  28  28  28  28  28  28  28   
29  29  29  29  29  29  29  29  29  29  29  ...  29  29  29  29  29  29  29   
..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ...  ..  ..  ..  ..  ..  ..  ..   
70  70  70  70  70  70  70  70  70  70  70  ...  70  70  70  70  70  70  70   
71  71  71  71  71  71  71  71  71  71  71  ...  71  71  71  71  71  71  71   
72  72  72  72  72  72  72  72  72  72  72  ...  72  72  72  72  72  72  72   
73  73  73  73  73  73  73  73  73  73  73  ...  73  73  73  73  73  73  73   
74  74  74  74  74  74  74  74  74  74  74  ...  74  74  74  74  74  74  74   
75  75  75  75  75  75  75  75  75  75  75  ...  75  75  75  75  75  75  75   
76  76  76  76  76  76  76  76  76  76  76  ...  76  76  76  76  76  76  76   
77  77  77  77  77  77  77  77  77  77  77  ...  77  77  77  77  77  77  77   
78  78  78  78  78  78  78  78  78  78  78  ...  78  78  78  78  78  78  78   
79  79  79  79  79  79  79  79  79  79  79  ...  79  79  79  79  79  79  79   
80  80  80  80  80  80  80  80  80  80  80  ...  80  80  80  80  80  80  80   
81  81  81  81  81  81  81  81  81  81  81  ...  81  81  81  81  81  81  81   
82  82  82  82  82  82  82  82  82  82  82  ...  82  82  82  82  82  82  82   
83  83  83  83  83  83  83  83  83  83  83  ...  83  83  83  83  83  83  83   
84  84  84  84  84  84  84  84  84  84  84  ...  84  84  84  84  84  84  84   
85  85  85  85  85  85  85  85  85  85  85  ...  85  85  85  85  85  85  85   
86  86  86  86  86  86  86  86  86  86  86  ...  86  86  86  86  86  86  86   
87  87  87  87  87  87  87  87  87  87  87  ...  87  87  87  87  87  87  87   
88  88  88  88  88  88  88  88  88  88  88  ...  88  8

The output of Chunk 1 and Chunk 2 is clearly not the same one, because a numpy.array follows row-major order,which means the base unit of a array is a row instead of a column, while a DataFrame is in column-major order.

In [19]:
m=100
n=100
k=100
A = np.array([i for i in range(m*n)]).reshape(m,n)
B = np.array([i for i in range(k*n)]).reshape(k,n)
# Method 1: Nested Loops
for i in range(m):
    for j in range(k):
        Sum = 0
        for l in range(n):
            Sum = Sum + A[i, l] / B[j, l]
        C[i, j] = Sum
C
# Method 2: np.dot()
print(np.dot(A,1/B.T))
# Method 3: Direct Computation
print(A/B)
# Method 4: Dimension Expansion
A = A.reshape(1, m, n)
B = B.reshape(k, 1, n)
C = A/B
np.sum(C, 2)

C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:11: RuntimeWarning: invalid value encountered in double_scalars
  # This is added back by InteractiveShellApp.init_path()


0       1       2       3       4       5       6       7       8  \
0      0.0     1.0     2.0     3.0     4.0     5.0     6.0     7.0     8.0   
1    100.0   101.0   102.0   103.0   104.0   105.0   106.0   107.0   108.0   
2    200.0   201.0   202.0   203.0   204.0   205.0   206.0   207.0   208.0   
3    300.0   301.0   302.0   303.0   304.0   305.0   306.0   307.0   308.0   
4    400.0   401.0   402.0   403.0   404.0   405.0   406.0   407.0   408.0   
5    500.0   501.0   502.0   503.0   504.0   505.0   506.0   507.0   508.0   
6    600.0   601.0   602.0   603.0   604.0   605.0   606.0   607.0   608.0   
7    700.0   701.0   702.0   703.0   704.0   705.0   706.0   707.0   708.0   
8    800.0   801.0   802.0   803.0   804.0   805.0   806.0   807.0   808.0   
9    900.0   901.0   902.0   903.0   904.0   905.0   906.0   907.0   908.0   
10  1000.0  1001.0  1002.0  1003.0  1004.0  1005.0  1006.0  1007.0  1008.0   
11  1100.0  1101.0  1102.0  1103.0  1104.0  1105.0  1106.0  1107.0  1108.0   
12  1200.0  1201.0  1202.0  1203.0  1204.0  1205.0  1206.0  1207.0  1208.0   
13  1300.0  1301.0  1302.0  1303.0  1304.0  1305.0  1306.0  1307.0  1308.0   
14  1400.0  1401.0  1402.0  1403.0  1404.0  1405.0  1406.0  1407.0  1408.0   
15  1500.0  1501.0  1502.0  1503.0  1504.0  1505.0  1506.0  1507.0  1508.0   
16  1600.0  1601.0  1602.0  1603.0  1604.0  1605.0  1606.0  1607.0  1608.0   
17  1700.0  1701.0  1702.0  1703.0  1704.0  1705.0  1706.0  1707.0  1708.0   
18  1800.0  1801.0  1802.0  1803.0  1804.0  1805.0  1806.0  1807.0  1808.0   
19  1900.0  1901.0  1902.0  1903.0  1904.0  1905.0  1906.0  1907.0  1908.0   
20  2000.0  2001.0  2002.0  2003.0  2004.0  2005.0  2006.0  2007.0  2008.0   
21  2100.0  2101.0  2102.0  2103.0  2104.0  2105.0  2106.0  2107.0  2108.0   
22  2200.0  2201.0  2202.0  2203.0  2204.0  2205.0  2206.0  2207.0  2208.0   
23  2300.0  2301.0  2302.0  2303.0  2304.0  2305.0  2306.0  2307.0  2308.0   
24  2400.0  2401.0  2402.0  2403.0  2404.0  2405.0  2406.0  2407.0  2408.0   
25  2500.0  2501.0  2502.0  2503.0  2504.0  2505.0  2506.0  2507.0  2508.0   
26  2600.0  2601.0  2602.0  2603.0  2604.0  2605.0  2606.0  2607.0  2608.0   
27  2700.0  2701.0  2702.0  2703.0  2704.0  2705.0  2706.0  2707.0  2708.0   
28  2800.0  2801.0  2802.0  2803.0  2804.0  2805.0  2806.0  2807.0  2808.0   
29  2900.0  2901.0  2902.0  2903.0  2904.0  2905.0  2906.0  2907.0  2908.0   
..     ...     ...     ...     ...     ...     ...     ...     ...     ...   
70  7000.0  7001.0  7002.0  7003.0  7004.0  7005.0  7006.0  7007.0  7008.0   
71  7100.0  7101.0  7102.0  7103.0  7104.0  7105.0  7106.0  7107.0  7108.0   
72  7200.0  7201.0  7202.0  7203.0  7204.0  7205.0  7206.0  7207.0  7208.0   
73  7300.0  7301.0  7302.0  7303.0  7304.0  7305.0  7306.0  7307.0  7308.0   
74  7400.0  7401.0  7402.0  7403.0  7404.0  7405.0  7406.0  7407.0  7408.0   
75  7500.0  7501.0  7502.0  7503.0  7504.0  7505.0  7506.0  7507.0  7508.0   
76  7600.0  7601.0  7602.0  7603.0  7604.0  7605.0  7606.0  7607.0  7608.0   
77  7700.0  7701.0  7702.0  7703.0  7704.0  7705.0  7706.0  7707.0  7708.0   
78  7800.0  7801.0  7802.0  7803.0  7804.0  7805.0  7806.0  7807.0  7808.0   
79  7900.0  7901.0  7902.0  7903.0  7904.0  7905.0  7906.0  7907.0  7908.0   
80  8000.0  8001.0  8002.0  8003.0  8004.0  8005.0  8006.0  8007.0  8008.0   
81  8100.0  8101.0  8102.0  8103.0  8104.0  8105.0  8106.0  8107.0  8108.0   
82  8200.0  8201.0  8202.0  8203.0  8204.0  8205.0  8206.0  8207.0  8208.0   
83  8300.0  8301.0  8302.0  8303.0  8304.0  8305.0  8306.0  8307.0  8308.0   
84  8400.0  8401.0  8402.0  8403.0  8404.0  8405.0  8406.0  8407.0  8408.0   
85  8500.0  8501.0  8502.0  8503.0  8504.0  8505.0  8506.0  8507.0  8508.0   
86  8600.0  8601.0  8602.0  8603.0  8604.0  8605.0  8606.0  8607.0  8608.0   
87  8700.0  8701.0  8702.0  8703.0  8704.0  8705.0  8706.0  8707.0  8708.0   
88  8800.0  8801.0  8802.0  8803.0  8804.0  8805.0  8806.0  8807.0  8808.0   
89  8900.0  8901.0  8902.0  890

C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:15: RuntimeWarning: divide by zero encountered in true_divide
  from ipykernel import kernelapp as app
C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:17: RuntimeWarning: invalid value encountered in true_divide


[[nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 ...
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]]
[[nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 ...
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]]


C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:21: RuntimeWarning: invalid value encountered in true_divide


array([[nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       ...,
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan]])

In [20]:
m=100
n=50
k=90
A = np.array([i for i in range(m*n)]).reshape(m,n)
B = np.array([i for i in range(k*n)]).reshape(k,n)
# Method 1: Nested Loops
start=time.clock()
for i in range(m):
    for j in range(k):
        Sum = 0
        for l in range(n):
            Sum = Sum + A[i, l] / B[j, l]
        C[i, j] = Sum
end=time.clock()
print("Method 1 costs",end-start)
# Method 2: np.dot()
start=time.clock()
print(np.dot(A,1/B.T))
end=time.clock()
print("Method 2 costs",end-start)
# Method 3: Direct Computation
# print(A/B)
# Method 4: Dimension Expansion
A = A.reshape(1, m, n)
B = B.reshape(k, 1, n)
C = A/B
np.sum(C, 2)

C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:7: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
  import sys
C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:12: RuntimeWarning: invalid value encountered in long_scalars
  if sys.path[0] == '':
C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:12: RuntimeWarning: divide by zero encountered in long_scalars
  if sys.path[0] == '':


Method 1 costs 1.3626389000000927
[[           nan 1.50913910e+01 9.28635957e+00 ... 2.79490916e-01
  2.76338482e-01 2.73256369e-01]
 [           inf 5.00000000e+01 2.96431798e+01 ... 8.50991021e-01
  8.41380090e-01 8.31983825e-01]
 [           inf 8.49086090e+01 5.00000000e+01 ... 1.42249112e+00
  1.40642170e+00 1.39071128e+00]
 ...
 [           inf 3.40122646e+03 1.98389792e+03 ... 5.57150010e+01
  5.50853745e+01 5.44698197e+01]
 [           inf 3.43613507e+03 2.00425474e+03 ... 5.62865011e+01
  5.56504161e+01 5.50285471e+01]
 [           inf 3.47104368e+03 2.02461156e+03 ... 5.68580013e+01
  5.62154577e+01 5.55872746e+01]]
Method 2 costs 0.005549499999688123


C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:14: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
  
C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:17: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: RuntimeWarning: divide by zero encountered in true_divide
C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:19: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:26: RuntimeWarning: divide by zero encountered in true_divide
C:\Users\rcf\Anaconda3\lib\site-packages\ipykernel_launcher.py:26: RuntimeWarning: inva

array([[           nan,            inf,            inf, ...,
                   inf,            inf,            inf],
       [1.50913910e+01, 5.00000000e+01, 8.49086090e+01, ...,
        3.40122646e+03, 3.43613507e+03, 3.47104368e+03],
       [9.28635957e+00, 2.96431798e+01, 5.00000000e+01, ...,
        1.98389792e+03, 2.00425474e+03, 2.02461156e+03],
       ...,
       [2.79490916e-01, 8.50991021e-01, 1.42249112e+00, ...,
        5.57150010e+01, 5.62865011e+01, 5.68580013e+01],
       [2.76338482e-01, 8.41380090e-01, 1.40642170e+00, ...,
        5.50853745e+01, 5.56504161e+01, 5.62154577e+01],
       [2.73256369e-01, 8.31983825e-01, 1.39071128e+00, ...,
        5.44698197e+01, 5.50285471e+01, 5.55872746e+01]])

Based on the time speed,Method 2 is the best choice.
Method 3 is wrong, because in numpy, C=A/B means C[i,j]=A[i,j]/B[i,j] instead of the normal matrix division.